### ISSUE TITLE
Annotate patchwork with line segments to create semantic plot relationships

### ISSUE THREAD

**Issue by mattwarkentin** [thomasp85/patchwork#216]
Hi @thomasp85,

I am wondering if it is currently possible to add line segments (such as those created with `geom_segment()`, for example) or any other `ggplot` annotations, on top of a patchwork to be able to create arbitrary relationships between plots. This could be useful for both standard patchworks or inset patchworks to create more obvious relationships between plots. Basically I'm trying to avoid doing any figure post-processing outside of `R`.

It is my understanding that the patchwork itself is a `ggplot` object, so I am wondering if this is possible, and if it isn't currently possible, if this feature is feasible.

Here is an illustrative example for what I'm hoping to achieve (I'm hoping to be able to add the dashed lines connecting the subplots in a patchwork):
<img width="825" alt="Screen Shot 2020-11-18 at 4 55 17 PM" src="https://user-images.githubusercontent.com/27825069/99592921-de000380-29be-11eb-91a4-8d39c5977549.png">

**Comment by thomasp85** [comment]
I agree that that would be very useful, but also non-trivial if at all possible. The reasons for this are pretty technical, but basically the different plots does not occupy a shared coordinate system and figuring out their relative locations are not straight-forward.

Don't hold your breath for it 🙂

**Comment by mattwarkentin** [comment]
I definitely do not have any of the technical knowledge to pretend to understand how `patchwork` or `ggplot2` work internally, but would it be possible to "flatten" or rasterize the entire patchwork and then treat that as a single grob upon which additional annotations could be plotted?

Obviously vector graphics are preferred for various reasons. I'm just thinking out loud.

**Comment by thomasp85** [comment]
Sure, you can just treat it as one big canvas - that is already possible. But there is no way to reference any existing information in the plot so you'd have to position the lines through trial and error. I'd advice just doing it in illustrator/inkscape instead

**Comment by mattwarkentin** [comment]
Congrats on the CRAN release, Thomas. To follow-up on your above comment, could you point me in the right direction for doing what you described above? I don't actually mind a bit of trial-and-error as long as the end result is reproducible. 

How does one "flatten" the patchwork into a single big canvas? I have tried using `annotate()` on a patchwork, but it seems to just add the annotation to the sub-plot canvases.

``` r
library(magrittr)
library(ggplot2)
library(patchwork)

p1 <- 
  mtcars %>% 
  ggplot(aes(mpg, hp)) +
  geom_point()

(p1 / p1) + annotate("text", 1, 1, label = "foo")
```

![](https://i.imgur.com/niXeFYS.png)

``` r
(p1 / p1) & annotate("text", 1, 1, label = "foo")
```

![](https://i.imgur.com/ZvzP2Nu.png)

**Comment by stoneyv** [comment]
@mattwarkentin There is a simple example at the end of this documentation that will yield annotations between ggplots grobs in different viewports in a grid using the package gggrid.
https://www.stat.auckland.ac.nz/~paul/Reports/gggrid/gggrid.html

gginards geom_debug(), ggpmisc
https://docs.r4photobiology.info/gginnards/
https://docs.r4photobiology.info/ggpmisc/
https://pparacch.github.io/2017/09/25/plotting_in_R_ggplot2_part_5.html
https://bookdown.org/rdpeng/RProgDA/the-grid-package.html

A great inspiration for visualizing relationships between data has been implemented by Susie Lu in d3-annotations d3js and react-annotations of the semiotic framework.
https://react-annotation.susielu.com/

**Comment by japhir** [comment]
I recently ran into a similar issue (at least I think?) where I'd like to draw lines between subplots in both the subplot's relevant reference frames.

For example:
```r
library(ggplot2)
library(patchwork)

# let's say we have two sites a and b, where each has its own "depth" scale:
dat <- tibble::tibble(depth = 1:200, a = rnorm(length(depth)), b = rnorm(length(depth)))

# we can make a nice patchwork showing  both (in this case the depth scale is identical but it could be very different)
(ggplot(dat) +
   aes(x = depth, y = a) + geom_line() + 
   # where these vertical lines show which parts are related to which part of the other site
   geom_vline(xintercept = c(0, 20, 30, 75, 150))) /
(ggplot(dat) +
   aes(x = depth, y = b) + geom_line() + 
   # so here we get different values
   geom_vline(xintercept = c(25, 50, 90, 120)))

ggsave("imgs/patchwork_annotation.png")
```
which results in the following image:

![patchwork_annotation](https://github.com/thomasp85/patchwork/assets/10659193/58ff96c9-877b-4c49-830e-6816c9542a68)

I'd like to be able to draw segments that connect these different elements in the different plots.
![patchwork_annotation_tweaked](https://github.com/thomasp85/patchwork/assets/10659193/c5386351-a2b7-4b42-a094-44c27f5d8057)

I think at present there is no way to draw these things that take into account the x-axis values from both subplots, right? And hacking stuff together with grids itself seems to prefer relative units? So far I've had to do this kind of thing in Inkscape, which is a bit of a hassle if I need to update my figures.
